File: src/main/java/com/example/homework/news/NewsPresenterImpl.java
package com.example.homework.news;

import android.support.annotation.NonNull;
import android.util.Log;

import com.example.homework.data.Translation;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class NewsPresenterImpl implements NewsContract.NewsPresenter {
    private NewsModel newsModel = new NewsModel();
    private NewsContract.NewsUIView newsUIView;

    public static String date;


    public NewsPresenterImpl(NewsContract.NewsUIView newsUIView) {
        this.newsUIView = newsUIView;
    }


    @Override
    public void initData() {
        Call<Translation> call = newsModel.update("latest");
        call.enqueue(new Callback<Translation>() {
            @Override
            public void onResponse(Call<Translation> call, Response<Translation> response) {
//                String responseData = response.body() != null ? response.body().toString() : null;
//                newsModel.parseJSONWithJSONObject(responseData);
                Translation translation = response.body();
                if (translation != null) {
                    date = translation.getDate();
                    newsUIView.initItems(translation);
                }
            }

            @Override
            public void onFailure(@NonNull Call<Translation> call, @NonNull Throwable t) {
                newsUIView.onError();
                t.printStackTrace();
            }
        });
//        newsModel.sendRequestWithOkHttp();

    }


    @Override
    public void refreshData() {
        newsModel.dataSet.clear();
        newsModel.topDataSet.clear();
        initData();
//        newsModel.sendRequestWithOkHttp();
        Log.v("refresh", "下拉刷新");
    }

    @Override
    public void onLoadMoreData() {
        Log.v("onLoadMore", "加载发送");
        Call<Translation> call = newsModel.update("before/" + date);
        call.enqueue(new Callback<Translation>() {
            @Override
            public void onResponse(Call<Translation> call, Response<Translation> response) {
//                String responseData = response.body() != null ? response.body().toString() : null;
//                newsModel.parseJSONWithJSONObject(responseData);
                Translation translation = response.body();
                if (translation != null) {
                    date = translation.getDate();
                    newsUIView.loadUpItems(translation);
                }
            }

            @Override
            public void onFailure(@NonNull Call<Translation> call, @NonNull Throwable t) {
                newsUIView.onError();
                t.printStackTrace();
            }
        });
//        newsModel.sendRequestWithOkHttp2();
//        newsUIView.loadUpItems(newsModel.dataSet);
    }
}

File: main/java/com/example/homework/news/NewsPresenterImpl.kt
package com.example.homework.news

import android.util.Log

import com.example.homework.data.Translation

import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class NewsPresenterImpl internal constructor(private val newsUIView: NewsContract.NewsUIView) : NewsContract.NewsPresenter {
    private val newsModel = NewsModel()

    var date: String? = null
    private var translation: Translation? = null


    override fun initData() {
        val call = newsModel.update("latest")
        call.enqueue(object : Callback<Translation> {
            override fun onResponse(call: Call<Translation>, response: Response<Translation>) {
                translation = response.body()
                if (translation != null) {
                    date = translation!!.date
                    newsUIView.initItems(translation!!)
                }
            }

            override fun onFailure(call: Call<Translation>, t: Throwable) {
                newsUIView.onError()
                t.printStackTrace()
            }
        })
    }

    override fun refreshData() {
        val call = newsModel.update("latest")
        call.enqueue(object : Callback<Translation> {
            override fun onResponse(call: Call<Translation>, response: Response<Translation>) {
                translation = response.body()
                if (translation != null) {
                    date = translation!!.date
                    newsUIView.refreshItems(translation!!)
                }
            }

            override fun onFailure(call: Call<Translation>, t: Throwable) {
                newsUIView.onError()
                t.printStackTrace()
            }
        })
        Log.v("refresh", "下拉刷新")
    }

    override fun onLoadMoreData() {
        val call = newsModel.update2(date)
        call.enqueue(object : Callback<Translation> {
            override fun onResponse(call: Call<Translation>, response: Response<Translation>) {
                Log.v("onLoadMore", "发请求了吗")
                translation = response.body()
                if (translation != null) {
                    date = translation!!.date
                    newsUIView.loadUpItems(translation!!)
                    Log.v("onLoadMore", "拿到东西了吗")
                }
            }

            override fun onFailure(call: Call<Translation>, t: Throwable) {
                newsUIView.onError()
                t.printStackTrace()
            }
        })
    }
}
